include/ginkgo/core/stop/residual_norm.hpp

```diff
@@ -312,6 +312,18 @@ class ImplicitResidualNorm : public ResidualNormBase<ValueType> {
 };
 
 
+// The following classes are deprecated, but they internally reference
+// themselves. To reduce unnecessary warnings, we disable deprecation warnings
+// for the definition of these classes.
+#if defined(__GNUC__) || defined(__clang__)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+#elif defined(_MSC_BUILD) || defined(__INTEL_LLVM_COMPILER)
+#pragma warning(push)
+#pragma warning(disable : 4996)
+#endif
```

Comment:
it only avoids the warning when we use it internally, but it still gives warning when users use it, right?

Reply:
Yes, this only disables the warning for the class definitions. After that, the previous warning state is restored.